### DISCUSSION TITLE
Question on @Riverpod(keepAlive: true) and ref.keepAlive()

### DISCUSSION THREAD

**Discussion by samleecampspot** [rrousselGit/riverpod#2561] [Q&A] [answered]
Hello!  

So I have an autoDispose FutureProvider that I would like to convert using the code-gen.  I had a `ref.keepAlive()` after an async call that can throw because I'd like to retry the call if it failed/threw before, but save the result when it succeeds.

```
final someProvider = FutureProvider.autoDispose<Object>((ref) async {
   final result = await someAsyncFunctionThatCanThrow();
   ref.keepAlive();
   return result;
});
```

If I wanted the same behavior, would it be okay to use the code-gen like this for the refactor?

```
@riverpod
Future<Object> someProvider(SomeProviderRef ref) async {
   final result = await someAsyncFunctionThatCanThrow();
   ref.keepAlive();
   return result;
}
```

or would `@Riverpod(keepAlive: true)` do what I need it to do?  I'm just worried it would save the error/throw state if I just used it like this:

```
@Riverpod(keepAlive: true)
Future<Object> someProvider(SomeProviderRef ref) async {
   final result = await someAsyncFunctionThatCanThrow();
   return result;
}
```

Any help here is appreciated; thanks!

**Comment by rrousselGit** [comment, marked as answer]
The annotation variant is equivalent to not using autoDispose.   
Using ref.keepAlive() is what you want in your specific case (of calling it after an `await`)

Generally:
- if your provider is _always_ kept alive, use the annotation parameter
- if it's only conditionally kept alive (such as after an await or fancier logic), then use `ref.keepAlive()`

**Reply by samleecampspot** [reply]
Thank you, that makes sense to me! 👍
